### ISSUE TITLE
Scanline count incorrect

### ISSUE THREAD

**Issue by thrust26** [stella-emu/stella#265]
I noticed this, because some PAL ROM which display color on TV show an odd scanline count in Stella (-> PAL color loss).

The title screen of Espial is one example. Stella 5.x counts 311 scanlines, but my PAL TV displays this in color. Stella 4.7.2 counts 310 scanlines and thus has no PAL color loss. Maybe this has to do with VSYNC not starting at the beginning of a scanline.

**Comment by DirtyHairy** [comment]
I just did a superficial test. Espial seems to strobe RSYNC every frame. If I disable rsync, I get 310 scanlines as in Stella 4.7.2, so I guess this is a scanline cut short by RSYNC, just like Fatal Run.

As for the others, I have no clue without a deeper analysis. @thrust26 , do I read you correctly that all of these show an odd number of scanlines in both Stella 4 and 5? Maybe there is more to PAL color loss thank just the "naked" scaline count.

**Comment by thrust26** [comment]
Artillery Duel shows an even number in 4.7.2 and an odd number in 5.x. That's like Espial.
The other ones show odd numbers in both versions.
And all show colors on a PAL TV.

Maybe there is more than one problem. Probably we should try to solve the problem why odd scanlines in Stella (both versions) show correctly on TV.

**Comment by thrust26** [comment]
I wrote a simple color loss test program and tested it on my own Sony Trinitron PAL TV. And it turns out, that it does **not care all** for odd or even scanlines! So at least my TV is no reference here.

@DirtyHairy @ale-79 and anyone with a PAL TV: Can you verify my results, please?
[ColorLoss.zip](https://github.com/stella-emu/stella/files/1542198/ColorLoss.zip)

Maybe there is no issue at all? Or maybe there even is no PAL color loss at all for a progressive signal?

**Comment by DirtyHairy** [comment]
@thrust26 I'll check it out on our LCD this weekend. My *guess* is that whether or not color information gets lost depends on the details and tolerance of the decoder circuit.

**Comment by thrust26** [comment]
I don't think any LCD will produce color loss under any conditions. We have to test this on CRTs.

**Comment by ale-79** [comment]
Currently I have a small 14" crt that I use for gaming. This TV shows color loss with odd scanline counts (confirmed with the test rom) just like Stella.
All the games in the first post with the exception of Espial display in B/W using this TV.
Espial is in color with no visible artifacts.

Some CRTs don't show color loss, but instead wrong colors in the top area of the screen. The height of the affected area seems to vary depending on the specific game, and it is also usually different every other frame, so that you can spot three different areas from top to bottom:
1- wrong colors (e.g in "artillery duel" the sky is magenta)
2- wrong and correct colors alternating each frame (the sky flickers between magenta and cyan - difficult to see in a still picture)
3- correct color (cyan sky)

Some colors seems to be unaffected (or very similar to the correct one), so the effect is not always so visible.


The following pictures are from an early 2000s 16/9 CRT showing this kind of artifact:
![01](https://user-images.githubusercontent.com/24704739/33778223-1b878308-dc48-11e7-8635-70487585b75b.JPG)
![02](https://user-images.githubusercontent.com/24704739/33778224-1f0b9096-dc48-11e7-8d4c-4c185f0d9e45.JPG)
![03](https://user-images.githubusercontent.com/24704739/33778229-227f415a-dc48-11e7-8aa9-388dd3dc3869.JPG)
![04](https://user-images.githubusercontent.com/24704739/33778231-24e5b67c-dc48-11e7-96dd-67cb37dae006.JPG)
![05](https://user-images.githubusercontent.com/24704739/33778233-270912f0-dc48-11e7-9a3a-6e658b1fb419.JPG)
![06](https://user-images.githubusercontent.com/24704739/33778236-298db666-dc48-11e7-90d5-65795c849215.JPG)

**Comment by sa666666** [comment]
I'm beginning to wonder if we can ever get this right without (as @DirtyHairy suggested at one point) emulating the TV colour decode circuitry.  Because what seems to be happening here is that colour loss is just one possible effect, and the naive approach that Stella uses (colour loss on odd scanline count) is simply inadequate and not always correct.

Sort of like the old TIA emulation, where we tried to emulate certain behaviour without understanding exactly where they came from.  When @DirtyHairy implemented the TIA as a combination of various counters, many other effect came 'for free'.  That is, once we emulated stuff at a lower level, all those strange artifacts just happened on their own, without emulating them specifically.

I've never really seen any documentation on how PAL TVs work.  Even the Blargg effects are technically NTSC only, and if we were being precise, they shouldn't even be used in PAL mode.  Maybe in the end, the TV effects (Blargg, etc) have to further developed to emulate the decode circuitry, and the colour-loss will come 'for free' too.

**Comment by sa666666** [comment]
Or to be more precise, we are currently trying to emulate the end result, when what is probably needed is emulation of the actual hardware at a lower level that is _causing_ this result.

**Comment by thrust26** [comment]
The color loss is coming from the eponymous Phase Alterning Line hardware. This means, the colors of two consecutive scanlines are transmitted with a phase shift of 180° for the **second** line and then averaged. Errors in the phase shift cause no hue error like in NTSC but only reduce the saturation, which is much less noticeable to the human eye. 

That's the part that I get so far. But how this causes a complete saturation loss when the scanline count is odd goes over my head. In my mind every 2nd frame would have the **first** scanline shifted by 180° and the 2nd not shifted. And then everything should work fine by averaging those. But obviously that's not how it works. Looks like I have to dig deeper.

Anyway, I think we have two unrelated issues here:
1. Espial (why is the scanline count wrong in 5.x?)
2. PAL color loss (this works as designed in Stella, except for the additional effects described OK)

IMO we should concentrate on fixing the first, because that's a bug and a regression. The second would be an enhancement.

**Comment by thrust26** [comment]
> I just did a superficial test. Espial seems to strobe RSYNC every frame. If I disable rsync, I get 310 scanlines as in Stella 4.7.2, so I guess this is a scanline cut short by RSYNC, just like Fatal Run.

So something with RSYNC emulation is not 100% correct, right? IIRC I once had the vague idea that RSYNC works differently for NTSC and PAL.

**Comment by DirtyHairy** [comment]
> So something with RSYNC emulation is not 100% correct, right? IIRC I once had the vague idea that RSYNC works differently for NTSC and PAL.

Check out #5 --- to me, @ale-79 results with the logic analyzers confirm RSYNC emulation. RSYNC effectively cuts shorts the scanline. It seems the old core didn't count the truncated line, while the old core does. The question to me is under which circumstances this additional line leads to color loss. I suppose it depends on the length of the resulting 'short line' and was going to write a test program to test this, which I never found time to to :smirk:

So, to sum it up, the emulation is fine, but we should refine line counting for those shortened lines. Or we leave it as is --- as @sa666666 says, our implementation of color loss is just an effective model anyway and will not describe the phenomenon 100% correct anyway.

**Comment by DirtyHairy** [comment]
>  But how this causes a complete saturation loss when the scanline count is odd goes over my head.

From my naive understanding, the decoding circuit has to phase shift the color signal before averaging, and the phase alternates between 0 and pi. Phase errors cancel out, while the signal (which was modulated with a corresponding phase shift) is preserved. Btw, humbuckers on an electric guitar work the same way 🎸 😏 

If the frame has odd scanline count, the phase of the first line will alternate between frames. I think that this somehow throws off the synchronization decoder circuit and messes up the relative phase, causing the color information to be erased by destructive interference.

**Comment by thrust26** [comment]
< So, to sum it up, the emulation is fine, but we should refine line counting for those shortened lines. Or we leave it as is --- as @sa666666 says, our implementation of color loss is just an effective model anyway and will not describe the phenomenon 100% correct anyway.

If the fix is not too complicated, I would opt for it. Probably that needs more research.

**Comment by sa666666** [comment]
The latest code in branch `feature/tv-emulation` partially addresses the RSYNC issue.  RSYNC does cut lines short, and counting scanlines for odd/even only works in certain cases.  If every scanline contributed the maximum number H_CLOCKS,  then that total and the number of scanlines being odd/even result in the same thing; colour loss.  However, RSYNC 'shorts' a line, so that line contributes fewer clocks that usual, but it still counts as a full line.  That's where we get different behaviour.

The current code in the branch now calculates the chroma phase, and whether is was inverted, based on clocks, and not on scanlines.  So that allows some PAL ROMs (Espial, Fatal Run) with odd scanline counts to show colour, since their cumulative cycle counts determine that the frame has the equivalent of an even number of scanlines, even when the scanline count is odd.

Again, I'm not 100% sure that this is what hardware does, but it does make sense to me.  And it does fit the findings above.  Comments are welcome, @DirtyHairy, @thrust26 and @ale-79.

**Comment by thrust26** [comment]
Sounds good to me. However, this seems like a more fundamental change. There are other parts in emulation which are based on scanlines. We should check them.

E.g. jitter emulation, which includes VSync lengths. Here is the code from the NTSC version of Fatal Run:
<img width="578" height="138" alt="Image" src="https://github.com/user-attachments/assets/54bc28ec-8dae-44e4-9adb-2dc7c05f3402" />
Due to the way RSYNC works, the VSync signal is only a bit over 2 scanlines long (worst case). While we expect at least 3 scanlines. But then, this is a prototype.

BTW: I still wonder why these games are doing that... 🤔 In Espial it is clearly a coding error, but in Fatal Run it looks intentional. Maybe it is a clever way to approximately start a new scanline when you are afraid that a WSYNC comes too late?

**Comment by sa666666** [comment]
Well technically, this is still sort of based on scanlines.  But the problem is scanlines is an int, and clocks can be a fraction.  So Fatal Run actually outputs 312.x scanlines.  The scanline counter records this as 313, but the PAL decoding circuitry seems to see it as 312.  Using total colour clocks instead of scanline count reconciles this difference.

I'm working on code now that will also show an effective emulation of colour loss for the entire frame, and also colour distortion as @ale-79 described.  But again, these are effective results, based on what we're seeing.  IOW, they're not necessarily a true/exact emulation of how the PAL signal works.

I'm also considering adding a mode where the PAL signal is reproduced exactly from first principles, from which all these effects will fall out naturally.  As I mentioned before, this is a rabbit hole, and I'm not sure how far I'd want to take it.  And the eventual end goal would likely to be do the same for NTSC too.

**Comment by thrust26** [comment]
Well, even if we emulate PAL or NTSC 100% correct, we can never 100% correctly emulate the output. Because every display reacts differently to signals. So we shouldn't waste too much time here. IMHO.

**Comment by sa666666** [comment]
Yes, the eventual result would still need knobs to adjust, just as a real TV would.  That's actually the easy part.  The harder part is getting the correct signal from the 2600 itself, and modelling how it deals with the TV.  Since it doesn't always react like later hardware, so some research will still be required.
